### PR TITLE
toolchain/binutils: switch to version 2.40 by default

### PR DESCRIPTION
**Pull request by hauke**
* toolchain/binutils: backport stable patches
    
    Add the patches with real changes from the binutils 2.40 stable branch.
    I am not aware that we ran into any of these problems, but I think it is
    better to take the existing stable patches.
    
    They were exported like this:
    git format-patch binutils-2_40...origin/binutils-2_40-branch
    I removed the patches changing the version numbers only.

*   toolchain/binutils: switch to version 2.40 by default
    
    This was build tested with all core packages on all targets
    successfully.
    
    This was run tested on the following systems:
   TODO

### CODE REVIEW

File: toolchain/binutils/patches/2.40/001-Re-enable-development.-Update-version-to-2.40.0.patch [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ delete mode 100644 gas/doc/.dirstamp
+
+--- a/bfd/development.sh
++++ b/bfd/development.sh
+@@ -16,7 +16,7 @@
+ # along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ 
+ # Controls whether to enable development-mode features by default.
+-development=false
++development=true
```
Comment (hauke) [inline toolchain/binutils/patches/2.40/001-Re-enable-development.-Update-version-to-2.40.0.patch:20]: I should probably remove this patch too.

Comment (lovehackintosh) [reply toolchain/binutils/patches/2.40/001-Re-enable-development.-Update-version-to-2.40.0.patch:20]: Yes, you should.

Comment (dhewg) [reply toolchain/binutils/patches/2.40/001-Re-enable-development.-Update-version-to-2.40.0.patch:20]: If follow up patches don't depend on it, the translation ones too, maybe?

**Comment by dhewg** [comment]
I've been using this on ramips and ipq4019 from day one, no issues related to that

**Comment by lovehackintosh** [comment]
You should also backport upstream commit 38e455b1d9a2a38d87a37180795c0872e08a9bf4

File: toolchain/binutils/patches/2.40/001-Re-enable-development.-Update-version-to-2.40.0.patch (hunk 1 above)
Comment (lovehackintosh) [reply toolchain/binutils/patches/2.40/001-Re-enable-development.-Update-version-to-2.40.0.patch:20]: @dhewg I don't think it is necessary.

**Comment by hauke** [comment]
The build of the ipq807x target failed for me like this:
````
====== Make errors from logs/package/kernel/qca-ssdk/compile.txt ======
Applying ./patches/0011-Revert-qca-ssdk-remove-bridge-fdb-entry-for-the-auth.patch using plaintext: 
patching file src/ref/ref_acl.c
make[4]: Entering directory '/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc'
mkdir -p /__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/build/bin
make -S MODULE_TYPE=KSLIB -C src/adpt all || exit 1;  make -S MODULE_TYPE=KSLIB -C src/ref all || exit 1;  make -S MODULE_TYPE=KSLIB -C src/sal all || exit 1;  make -S MODULE_TYPE=KSLIB -C src/api all || exit 1;  make -S MODULE_TYPE=KSLIB -C src/util all || exit 1;  make -S MODULE_TYPE=KSLIB -C src/fal all || exit 1;  make -S MODULE_TYPE=KSLIB -C src/hsl all || exit 1;  make -S MODULE_TYPE=KSLIB -C src/shell_lib all || exit 1;  make -S MODULE_TYPE=KSLIB -C src/init all || exit 1;
make[5]: Entering directory '/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/src/adpt'
mkdir -p /__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/build/linux/KSLIB
ccache_cc -Os -pipe -mcpu=cortex-a53 -fno-caller-saves -fno-plt  -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro  -I/__w/openwrt/openwrt/openwrt/openwrt-toolchain-ipq807x-generic_gcc-12.2.0_musl.Linux-x86_64/toolchain-aarch64_cortex-a53_gcc-12.2.0_musl/usr/include -I/__w/openwrt/openwrt/openwrt/openwrt-toolchain-ipq807x-generic_gcc-12.2.0_musl.Linux-x86_64/toolchain-aarch64_cortex-a53_gcc-12.2.0_musl/include/fortify -I/__w/openwrt/openwrt/openwrt/openwrt-toolchain-ipq807x-generic_gcc-12.2.0_musl.Linux-x86_64/toolchain-aarch64_cortex-a53_gcc-12.2.0_musl/include   -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/common -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/api -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/fal -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/ref -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/adpt -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/hsl -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/hsl/phy -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/sal/os -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/sal/os/linux -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/sal/sd -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/sal/sd/linux/hydra_howl -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/sal/sd/linux/uk_interface -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/init -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/hsl/isisc -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/hsl/hppe -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/adpt/hppe -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/hsl/mht -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/hsl/sfp -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/adpt/sfp -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/include/shell_lib -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/linux-5.15.94 -I./usr/bin ./bin/../lib/gcc/aarch64-openwrt-linux-musl/120200/include/ -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/linux-5.15.94/include -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/linux-5.15.94/source -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/linux-5.15.94/source/include -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/linux-5.15.94/source/arch/arm64/mach-msm/include -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/linux-5.15.94/arch/arm64/mach-msm/include -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/linux-5.15.94/source/arch/arm64/include -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/linux-5.15.94/arch/arm64/include -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/linux-5.15.94/source/arch/arm64/include/asm -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/linux-5.15.94/arch/arm64/include/generated -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/linux-5.15.94/include/generated/uapi -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/linux-5.15.94/include/uapi -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/linux-5.15.94/arch/arm64/include/uapi -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/linux-5.15.94/source/arch/arm64/include/asm/mach -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/linux-5.15.94/arch/arm64/include/generated/uapi -I/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/linux-5.15.94/source/include/uapi -include /__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/linux-5.15.94/include/linux/kconfig.h  -DIN_ACL -DIN_FDB -DIN_IGMP -DIN_LEAKY -DIN_LED -DIN_MIB -DIN_MIRROR -DIN_MISC -DIN_PORTCONTROL -DIN_PORTVLAN -DIN_QOS -DIN_RATE -DIN_STP -DIN_VLAN -DIN_COSMAP -DIN_IP -DIN_NAT -DIN_FLOW -DIN_TRUNK -DIN_SEC -DIN_QM -DIN_INTERFACECONTROL -DIN_CTRLPKT -DIN_SERVCODE -DIN_RSS_HASH -DIN_MALIBU_PHY -DIN_AQUANTIA_PHY -DIN_QCA803X_PHY -DIN_QCA808X_PHY -DIN_SFP_PHY -DIN_PHY_I2C_MODE -DIN_VSI -DIN_PPPOE -DIN_BM -DIN_SHAPER -DIN_POLICER -DIN_UNIPHY -DUK_IF -DUK_NL_PROT= -DUK_MINOR_DEV=254 -DIN_NETLINK  -Wall -DVERSION=\"3.0.0.2\" -DBUILD_DATE=\"2023-02-21-00:51:44\" -DOS=\"linux\" -D"KBUILD_STR(s)=#s" -D"KBUILD_MODNAME=KBUILD_STR(qca-ssdk)" -DISISC -DHPPE -DMHT -DIN_SFP -DKVER34 -DKVER32 -DLNX26_22 -D__KERNEL__ -DKERNEL_MODULE -DMODULE -Os -pipe -march=armv8-a -mcpu=cortex-a53+crypto -fno-caller-saves -fno-strict-aliasing -fno-common -Wno-format-security -Wno-pointer-sign -Wno-unused-but-set-variable -Wno-error=unused-result -mcmodel=large -fno-stack-protector -MM adpt.c > /__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/build/linux/KSLIB/adpt.d.tmp
aarch64-openwrt-linux-musl-gcc.bin: warning: ./bin/../lib/gcc/aarch64-openwrt-linux-musl/120200/include/: linker input file unused because linking not done
aarch64-openwrt-linux-musl-gcc.bin: error: ./bin/../lib/gcc/aarch64-openwrt-linux-musl/120200/include/: linker input file not found: No such file or directory
make[5]: *** [/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/make/target.mk:16: adpt.d] Error 1
make[5]: Leaving directory '/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/src/adpt'
make[4]: *** [Makefile:27: kslib_o] Error 1
make[4]: Leaving directory '/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc'
make[3]: *** [Makefile:85: /__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-ipq807x_generic/qca-ssdk-2022-09-12-628b22bc/.built] Error 2
time: package/kernel/qca-ssdk/compile#0.42#0.16#0.85
Error: Process completed with exit code 2.
````

**Comment by robimarko** [comment]
@hauke I just tried and SSDK built for me with binutils 2.40

**Comment by hauke** [comment]
@robimarko  This the ipq807x build problem was in my test build with the default toolchain. 
The results are here: https://github.com/hauke/openwrt/actions/runs/4227308102/jobs/7341671088

There is probably a bug in the build dependencies in this application. This is not related to binutils 2.40.

**Comment by robimarko** [comment]
I had some random reports of failures like that, but then it just works.
For me it always builds, CI and buildbots also dont have issues, but yes SSDK is a terrible piece of SW.

**Pull request merged**